### PR TITLE
[Website] Add last modified header

### PR DESCRIPTION
**Pull request by drewfreyling**

### CODE REVIEW

**Comment by drewfreyling** [comment]
Agreed.

**Comment by PeterDaveHello** [comment]
The tricky point is the header for library page, we don't have timestamp on library update yet.

**Comment by drewfreyling** [comment]
We can leave that page off the 304 list for now until that is in place.

**Comment by drewfreyling** [comment]
@PeterDaveHello i've removed the extra files and removed the libraries page from the cache validity check.

**Comment by hhmmrrhh** [comment]
Ihgi ioioy n0y uoj9u9ytt oho

**Comment by MattIPv4** [comment]
@drewfreyling Hey 👋 Could you please resolve the conflicts and then I can chase this being merged?

**Comment by drewfreyling** [comment]
I'll get it fixed it up.

**Comment by MattIPv4** [comment]
@PeterDaveHello Can you review this please?

File: webServer.js [hunk 1]
```diff
@@ -124,8 +132,8 @@ function start() {
     error: getTemplate('templates/500.html', true)
   };
 
-  var generatePage = function (options) {
-    var layout = options.layout || templates.layout;
+  var generatePage = function(options) {
+    var layout = options.layout.content | templates.layout.content;
```
Comment (MattIPv4) [inline webServer.js]: Shouldn't this be `||`?
I don't think that bitwise operations on strings are advised..?

File: webServer.js [hunk 2]
```diff
@@ -124,8 +132,8 @@ function start() {
     error: getTemplate('templates/500.html', true)
   };
 
-  var generatePage = function (options) {
-    var layout = options.layout || templates.layout;
+  var generatePage = function(options) {
```
Comment (MattIPv4) [inline webServer.js]: Can this space be restored please :)

File: webServer.js [hunk 3]
```diff
@@ -146,9 +154,29 @@ function start() {
     return fullContent;
   };
 
-  var setCache = function (res, hours) {
-    res.setHeader('Cache-Control', 'public, max-age=' + 60 * 60 * hours + ', immutable');
-    res.setHeader('Expires', new Date(Date.now() + 60 * 60 * hours * 1000).toUTCString());
+  var cacheCheck = function(req, res, hours, lastModified) {
```
Comment (MattIPv4) [inline webServer.js]: Can there be a space between `function` and `(`?

File: webServer.js [hunk 4]
```diff
@@ -446,46 +479,49 @@ function start() {
...
-    res.send(generatePage({
-      reqUrl: req.url,
-      title: 'libraries - ' + TITLE,
-      page: {
-        template: templates.libraries,
-        data: {
-          packages: _.toArray(LIBRARIES_MAP),
-          libCount: Object.keys(LIBRARIES_MAP).length,
-          libVerCount: LIBRARIES_VERSIONS
+  app.get('/libraries', function(req, res) {
```
Comment (MattIPv4) [inline webServer.js]: Space between `function` & `(` please.

File: webServer.js [hunk 5]
```diff
@@ -496,26 +532,30 @@ function start() {
...
-  app.get('/about', function (req, res) {
-    setCache(res, 72);
-    res.send(generatePage({
-      reqUrl: req.url,
-      title: 'about - ' + TITLE,
-      page: {
-        template: templates.about
-      }
-    }));
+  app.get('/about', function(req, res) {
```
Comment (MattIPv4) [inline webServer.js]: Space in function declaration, please.

File: webServer.js [hunk 6]
```diff
@@ -496,26 +532,30 @@ function start() {
...
-  app.get('/api', function (req, res) {
-    setCache(res, 72);
-    res.send(generatePage({
-      reqUrl: req.url,
-      title: 'API - ' + TITLE,
-      page: {
-        template: templates.api
-      }
-    }));
+  app.get('/api', function(req, res) {
```
Comment (MattIPv4) [inline webServer.js]: Space once again :)

File: webServer.js (hunk 4 above)
Comment (drewfreyling) [reply webServer.js]: We should probably add an `.eslintrc` file and set https://eslint.org/docs/rules/space-before-function-paren

File: webServer.js (hunk 1 above)
Comment (drewfreyling) [reply webServer.js]: I'll fix.

File: webServer.js (hunk 4 above)
Comment (PeterDaveHello) [reply webServer.js]: Good idea!

**Comment by PeterDaveHello** [comment]
Will review again once above issues resolved ;)

File: webServer.js (hunk 4 above)
Comment (drewfreyling) [reply webServer.js]: I'll fix these up and submit a pr for that later.

Comment (drewfreyling) [reply webServer.js]: https://github.com/cdnjs/new-website/pull/280 fixes these types of things going forward.

File: webServer.js [hunk 7]
```diff
@@ -146,9 +154,29 @@ function start() {
...
+
+    //Get the if-modified-since header from the request
+    var reqModDate = req.headers["if-modified-since"];
+
+    //check if if-modified-since header is the same as the mtime of the file
+    if (lastModified && reqModDate != null) {
+      reqModDate = new Date(reqModDate);
+
+      //compared dates to the nearest second as that is what the if modified since header is accurate to
+      if (Math.floor(reqModDate.getTime() / 1000) == Math.floor(lastModified.getTime() / 1000)) {
```
Comment (MattIPv4) [inline webServer.js]: Should we use strict comparison here, or is normal equals (as it is) fine?

File: webServer.js [hunk 8]
```diff
@@ -146,9 +154,29 @@ function start() {
     return fullContent;
   };
 
-  var setCache = function (res, hours) {
-    res.setHeader('Cache-Control', 'public, max-age=' + 60 * 60 * hours + ', immutable');
-    res.setHeader('Expires', new Date(Date.now() + 60 * 60 * hours * 1000).toUTCString());
+  var cacheCheck = function (req, res, hours, lastModified) {
+    res.setHeader("Cache-Control", "public, max-age=" + 60 * 60 * hours + ", immutable");
```
Comment (MattIPv4) [inline webServer.js]: Should .toString() be used on the number in this line, or is the type coercion that is currently happening fine?

Comment (drewfreyling) [reply webServer.js]: I think it's fine.

File: webServer.js (hunk 7 above)
Comment (drewfreyling) [reply webServer.js]: Changed to triple equals.

**Comment by MattIPv4** [comment]
cc #157

**Comment by drewfreyling** [comment]
> Something broken?
> 
> ```
> TypeError: Cannot read property 'content' of undefined
>     at generatePage (/home/peter/new-website/webServer.js:136:33)
>     at /home/peter/new-website/webServer.js:573:26
>     at Layer.handle_error (/home/peter/new-website/node_modules/express/lib/router/layer.js:71:5)
>     at trim_prefix (/home/peter/new-website/node_modules/express/lib/router/index.js:315:13)
>     at /home/peter/new-website/node_modules/express/lib/router/index.js:284:7
>     at Function.process_params (/home/peter/new-website/node_modules/express/lib/router/index.js:335:12)
>     at next (/home/peter/new-website/node_modules/express/lib/router/index.js:275:10)
>     at Layer.handle_error (/home/peter/new-website/node_modules/express/lib/router/layer.js:67:12)
>     at trim_prefix (/home/peter/new-website/node_modules/express/lib/router/index.js:315:13)
>     at /home/peter/new-website/node_modules/express/lib/router/index.js:284:7
> ```

It might be from the rebase, I'll investigate.

**Comment by MattIPv4** [comment]
@PeterDaveHello I believe that error is now fixed, I cannot reproduce it when navigating about the site.

**Comment by drewfreyling** [comment]
@MattIPv4 - I'm assuming @PeterDaveHello  won't be reviewing this?

**Comment by MattIPv4** [comment]
I would assume the same. I'll pull this down in a bit and check nothing is broken still.

**Comment by PeterDaveHello** [comment]
@drewfreyling I've increased the permission level, you should be able to help review and merge pull requests now

**Pull request merged**